src/query/storage/m3/cluster_resolver.go

```diff
@@ -0,0 +1,293 @@
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3
+
+import (
+	"sort"
+	"time"
+
+	"github.com/m3db/m3/src/query/storage"
+)
+
+type unaggregatedNamespaceType uint8
+
+const (
+	partiallySatisfiesRange unaggregatedNamespaceType = iota
+	fullySatisfiesRange
+	disabled
+)
+
+type unaggregatedNamespaceDetails struct {
+	satisfies        unaggregatedNamespaceType
+	clusterNamespace ClusterNamespace
+	retention        time.Duration
+}
+
+// resolveUnaggregatedNamespaceForQuery determines if the unaggregated namespace
+// should be used, and if so, determines if it fully satisfies the query range.
+func resolveUnaggregatedNamespaceForQuery(
+	now time.Time,
+	start time.Time,
+	unaggregated ClusterNamespace,
+	opts *storage.FanoutOptions,
+) unaggregatedNamespaceDetails {
+	if opts.FanoutUnaggregated == storage.FanoutForceDisable {
+		return unaggregatedNamespaceDetails{satisfies: disabled}
+	}
+
+	retention := unaggregated.Options().Attributes().Retention
+	unaggregatedStart := now.Add(-1 * retention)
+	if unaggregatedStart.Before(start) || unaggregatedStart.Equal(start) {
+		return unaggregatedNamespaceDetails{
+			clusterNamespace: unaggregated,
+			retention:        retention,
+			satisfies:        fullySatisfiesRange,
+		}
+	}
+
+	return unaggregatedNamespaceDetails{
+		clusterNamespace: unaggregated,
+		retention:        retention,
+		satisfies:        partiallySatisfiesRange,
+	}
+}
+
+// resolveClusterNamespacesForQuery returns the namespaces that need to be
+// fanned out to depending on the query time and the namespaces configured.
+func resolveClusterNamespacesForQuery(
+	now time.Time,
+	clusters Clusters,
+	start time.Time,
+	end time.Time,
+	opts *storage.FanoutOptions,
+) (queryFanoutType, ClusterNamespaces, error) {
+	// First check if the unaggregated cluster can fully satisfy the query range.
+	// If so, return it and shortcircuit, as unaggregated will necessarily have
+	// every metric.
+	unaggregated := resolveUnaggregatedNamespaceForQuery(now, start,
+		clusters.UnaggregatedClusterNamespace(), opts)
+	if unaggregated.satisfies == fullySatisfiesRange {
+		return namespaceCoversAllQueryRange,
+			ClusterNamespaces{unaggregated.clusterNamespace},
+			nil
+	}
+
+	if opts.FanoutAggregated == storage.FanoutForceDisable {
+		if unaggregated.satisfies == partiallySatisfiesRange {
+			return namespaceCoversAllQueryRange,
```

Comment:
nit: Probably better to return `namespaceCoversPartialQueryRange` here since it only partially satisfies the range?